src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,630 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+package sun.font;
+
+import java.awt.geom.Point2D;
+import sun.font.GlyphLayout.GVData;
+import sun.java2d.Disposer;
+import sun.java2d.DisposerRecord;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import static java.lang.foreign.MemorySegment.NULL;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.SymbolLookup;
+import java.lang.foreign.UnionLayout;
+import static java.lang.foreign.ValueLayout.*;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+
+import java.util.Optional;
+import java.util.WeakHashMap;
+
+public class HBShaper {
+
+    /*
+     * union _hb_var_int_t {
+     *     uint32_t u32;
+     *     int32_t i32;
+     *     uint16_t u16[2];
+     *     int16_t i16[2];
+     *     uint8_t u8[4];
+     *     int8_t i8[4];
+     * };
+     */
+   private static final UnionLayout VarIntLayout = MemoryLayout.unionLayout(
+        JAVA_INT.withName("u32"),
+        JAVA_INT.withName("i32"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("u16"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("i16"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("u8"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("i8")
+    ).withName("_hb_var_int_t");
+
+    /*
+     * struct hb_glyph_position_t {
+     *     hb_position_t x_advance;
+     *     hb_position_t y_advance;
+     *     hb_position_t x_offset;
+     *     hb_position_t y_offset;
+     *     hb_var_int_t var;
+     * };
+     */
+     private static final StructLayout PositionLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("x_advance"),
+        JAVA_INT.withName("y_advance"),
+        JAVA_INT.withName("x_offset"),
+        JAVA_INT.withName("y_offset"),
+        VarIntLayout.withName("var")
+     ).withName("hb_glyph_position_t");
+
+    /**
+     * struct hb_glyph_info_t {
+     *     hb_codepoint_t codepoint;
+     *     hb_mask_t mask;
+     *     uint32_t cluster;
+     *     hb_var_int_t var1;
+     *     hb_var_int_t var2;
+     * };
+     */
+    private static final StructLayout GlyphInfoLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("codepoint"),
+        JAVA_INT.withName("mask"),
+        JAVA_INT.withName("cluster"),
+        VarIntLayout.withName("var1"),
+        VarIntLayout.withName("var2")
+    ).withName("hb_glyph_info_t");
+
+    private static VarHandle getVarHandle(MemoryLayout layout, String name) {
```

Comment:
This method could take a `SequenceLayout` instead of a `MemoryLayout` as it only works for SeequenceLayouts.